libs/ktxreader/src/Ktx2Reader.cpp

```diff
@@ -136,15 +223,99 @@ void Ktx2Reader::unrequestFormat(Texture::InternalFormat format) {
     }
 }
 
-Texture* Ktx2Reader::load(const uint8_t* data, size_t size, TransferFunction transfer) {
-    if (!mTranscoder->init(data, size)) {
+Texture* Ktx2Reader::load(const void* data, size_t size, TransferFunction transfer) {
+    Texture* texture = createTexture(mTranscoder.get(), data, size, transfer);
+    if (texture == nullptr) {
+        return nullptr;
+    }
+
+    if (!mTranscoder->start_transcoding()) {
+        mEngine.destroy(texture);
+        if (!mQuiet) {
+            utils::slog.e << "BasisU start_transcoding failed." << utils::io::endl;
+        }
+        return nullptr;
+    }
+
+    ktx2_transcoder_state basisThreadState;
+    basisThreadState.clear();
+
+    for (uint32_t levelIndex = 0; levelIndex < mTranscoder->get_levels(); levelIndex++) {
+        Texture::PixelBufferDescriptor* pbd;
+        Result result = transcodeImageLevel(*mTranscoder, basisThreadState, texture->getFormat(),
+                levelIndex, &pbd);
+        if (result != Result::SUCCESS) {
+            mEngine.destroy(texture);
+            utils::slog.e << "Failed to transcode level " << levelIndex << utils::io::endl;
+            return nullptr;
+        }
+        texture->setImage(mEngine, levelIndex, std::move(*pbd));
+    }
+    return texture;
+}
+
+Result FAsync::doTranscoding() {
+    ktx2_transcoder_state basisThreadState;
+    basisThreadState.clear();
+    for (uint32_t levelIndex = 0; levelIndex < mTranscoder->get_levels(); levelIndex++) {
+        Texture::PixelBufferDescriptor* pbd;
+        Result result = transcodeImageLevel(*mTranscoder, basisThreadState, mTexture->getFormat(),
+                levelIndex, &pbd);
+        if (result != Result::SUCCESS) {
+            return result;
+        }
+        mTranscoderResults[levelIndex].store(pbd);
+    }
+    return Result::SUCCESS;
+}
+
+void FAsync::uploadImages() {
+    size_t levelIndex = 0;
+    for (TranscoderResult& level : mTranscoderResults) {
```

Comment:
I suggest `UTILS_NOUNROLL` here because the compiler usually compeltely unrolls loops when the count is known at compile time.